internal/spokes/spokes.go

```diff
@@ -94,12 +94,20 @@ func Exec(ctx context.Context, stdin io.Reader, stdout io.Writer, stderr io.Writ
 
 	if err := rp.execute(ctx); err != nil {
 		g.SetError(1, err.Error())
+		rp.RemoveQuarantine()
 		return 1, fmt.Errorf("unexpected error running spokes receive pack: %w", err)
 	}
 
 	return 0, nil
 }
 
+func (r *spokesReceivePack) cleanQuarantineOnContextDone(ctx context.Context) {
+	go func() {
+		<-ctx.Done()
```

Comment:
The ctx.Done often is closed on both success and failure, so this could end up removing the quarantine on success. Also, if the context was canceled and interrupted something, the interrupted code is responsible for returning an error.

I would suggest removing this method and only calling removequarantine when rp.execute returns an error.

Reply:
I have added bda519dc148eb7b4ecfc5922d87e455754a7787f where only `rp.execute` is checked to return an error.

> Also, if the context was canceled and interrupted something, the interrupted code is responsible for returning an error.

I don't think this is going to cover all the potential error paths, so, for example, if `spokes-receive-pack` gets terminated while we are executing the `report`:

```go
if capabilities.IsDefined(pktline.ReportStatusV2) || capabilities.IsDefined(pktline.ReportStatus) {
		if err := r.report(ctx, unpackErr == nil, commands, capabilities); err != nil {
			return err
		}
	}
```

our actual error handling code won't be able to deal with the signal, and the quarantine folder won't be cleaned.

Reply:
Yes, that's true, `Read` and `Write` don't react to context cancelation or to signals, so those could stall. The simplest way to deal with this is to do `if ctx.Err() != nil { return ctx.Err() }` before every Read or Write call, though that doesn't ensure the calls themselves abort on cancelation. It is possible to do reads and writes asynchronously so that you can also check for context cancelation and abort immediately in those cases.

I suspect that spokes-receive-pack is killed by a signal very rarely, so there will be very few quarantines that leak if the signal handling isn't perfect here.

In the past week, `spokesctl gov aggregate die_message count -P spokes-receive-pack` reports 5 exits that look related to signals vs ~500k exits due to a problem that index-pack detected. Of the 500k, gitmon shows that index-pack reported 200k errors because the client disconnected, and the rest were due to other data errors. (I didn't paste the actual output here because this is a public repo.) The disconnects will show up as stdin and/or stdout getting closed, which will make read and write abort, which will lead to the cleanup happening.

So this branch should be a big improvement even if read and write don't get aborted correctly in every case.